lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -453,6 +460,68 @@ void SPIRVRegularizeLLVMBase::lowerFuncPtr(Module *M) {
     lowerFuncPtr(I.first, I.second);
 }
 
+void SPIRVRegularizeLLVMBase::addKernelEntryPoint(Module *M) {
+  std::vector<Function *> Work;
+
+  // Get a list of all functions that have SPIR kernel calling conv
+  for (auto &F : *M) {
+    if (F.getCallingConv() != CallingConv::SPIR_KERNEL)
+      continue;
+
+    Work.push_back(&F);
+  }
+  for (auto &F: Work) {
+
+    // for declarations just make them into SPIR functions.
+    if (F->empty()) {
+      F->setCallingConv(CallingConv::SPIR_FUNC);
+      continue;
+    }
+
+    // Otherwise add a wrapper around the function to act as an entry point.
+    FunctionType *FType = F->getFunctionType();
+    std::string wrap_name = "__spirv_entry_" + static_cast<std::string>(F->getName());
+    Function *WrapFn = getOrCreateFunction(M, F->getReturnType(),
+					   FType->params(), wrap_name);
+
+    auto *CallBB = BasicBlock::Create(M->getContext(), "", WrapFn);
+    IRBuilder<> Builder(CallBB);
+
+    SmallVector<Value *, 1> Args;
+    for (auto AI = WrapFn->arg_begin(), End = WrapFn->arg_end(); AI != End; ++AI) {
+      Value *A = AI;
+      Args.emplace_back(A);
+    }
+    F->setCallingConv(CallingConv::SPIR_FUNC);
+    auto CI = CallInst::Create(F, ArrayRef<Value *>(Args), F->getReturnType()->isVoidTy() ? "" : "call", CallBB);
+    CI->setCallingConv(F->getCallingConv());
```

Comment:
I suggest to try [CloneFunction](https://llvm.org/doxygen/CloneFunction_8cpp_source.html#l00283)

Reply:
I don't think CloneFunction does what I want here, it seems to be for cloning functions between modules, I'm only cloning the decl and adding a call.